core/templates/dev/head/pages/exploration_editor/editor_tab/SolutionVerificationServiceSpec.js

```diff
@@ -0,0 +1,199 @@
+              correct: false,
+              outcome: {
+                dest: 'End State',
+                feedback: [],
+                param_changes: []
+              },
+              rule_specs: [{
+                inputs: {x: 'abc'},
+                rule_type: 'Contains'
+              }]
+            }],
+            default_outcome: {
+              dest: 'default',
+              feedback: [],
+              param_changes: []
+            },
+            fallbacks: [],
+            hints: [{
+              hint_text: 'one'
+            }, {
+              hint_text: 'two'
+            }]
+          },
+          param_changes: []
+        },
+        'End State': {
+          content: {
+            html: '',
+            audio_translations: {}
+          },
+          interaction: {
+            id: 'TextInput',
+            answer_groups: [{
+              rule_specs: [],
+              outcome: {},
+              correct: false
+            }],
+            default_outcome: {
+              dest: 'default',
+              feedback: [],
+              param_changes: []
+            },
+            fallbacks: [],
+            hints: []
+          },
+          param_changes: []
+        }
+      });
+
+      var stateEditorCtrl = $controller('StateEditor', {
+        $scope: scope,
+        editorContextService: ecs,
+        changeListService: cls,
+        explorationStatesService: ess,
+        editabilityService: {
+          isEditable: function() {
+            return true;
+          }
+        },
+        INTERACTION_SPECS: IS
+      });
+
+      var interactionCtrl = $controller('StateInteraction', {
+        $scope: scope,
+        editorContextService: ecs,
+        changeListService: cls,
+        explorationStatesService: ess,
+        editabilityService: {
+          isEditable: function() {
+            return true;
+          }
+        },
+        stateInteractionIdService: siis,
+        stateCustomizationArgsService: scas,
+        interactionDetailsCache: idc,
+        INTERACTION_SPECS: IS
+      });
+    }));
+
+    it('should verify the solution correctly',
+      function() {
+        ecs.setActiveStateName('First State');
+        scope.initStateEditor();
+
+        var state = ess.getState('First State');
+        siis.init(
+          'First State', state.interaction.id, state.interaction, 'widget_id');
+        scas.init(
+          'First State', state.interaction.customizationArgs,
+          state.interaction, 'widget_customization_args');
+
+        siis.displayed = 'TextInput';
+        scope.onCustomizationModalSavePostHook();
+
+        ess.saveSolution('First State', sof.createNew(false, 'abc', 'nothing'));
+
+        svs.verifySolution(0, state,
+          ess.getState('First State').interaction.solution.correctAnswer,
+          function() {
+            ess.updateSolutionValidity('First State', true);
+            expect(ess.isSolutionValid('First State')).toBe(true);
```

Comment:
Shouldn't these expectations (this one and the one in line 194) come at the very end of the whole test? Otherwise, if this function is not called, the expectation will not trigger.

If you are concerned about the async operation, see done(): https://jasmine.github.io/2.0/introduction.html#section-Asynchronous_Support